### PR TITLE
Rename things to match protocol specification

### LINKED ISSUE #602: Rename code around addresses, keys, notes, nullifiers, encryption/decryption, and spend authority.
Change type/variable names in code to match the following naming conventions (which were cut'n'pasted from #324, a ticket which mixed together multiple issues):

# Dev terminology

Let's begin distinguishing these three items:

* `bcaddr` and `bckey` for upstream "cleartext' Bitcoin address/key pairs.
* `zc-spend-authority address / key` for the second category. (It's unwieldy and only needs to be used in specialized topics.)
* `zc-transmission address / key` for the third category. (Also unwieldy and specialized.)
* `zcaddr` and `zckey` for the *combined* spend-and-transmit values.

So for example, at some layer of abstraction, the parameters to construct a Pour are `zcaddr` and `zckey` types, rather than the more distinct keys used behind that abstraction.

### PR DESCRIPTION
**Pull request by ebfull**
This is a pass through the code to eliminate usage of "pour" and "serial" and any other outdated terminology.

Closes #602

### CODE REVIEW

**Comment by bitcartel** [comment]
I don't want to reopen zcash/zips#26 but how about 'stub' instead of 'nullifier'?  So we have a collection of stubs (like a check book) where a stub provides a record of a transaction and its counterpart, the check, can only be used once.

**Comment by ebfull** [comment]
Although that's a good suggestion, I still like nullifier better. :sunglasses:

**Comment by jl777** [comment]
"stub" has a specific meaning that would add confusion
but nullifier does sound a bit unusual, and its literal meaning "makes everything zero" does not fit what it does "marks an output as spent"

**Comment by zookozcash** [comment]
I like the idea of 'stub', but is too late to change terminology.

**Comment by daira** [comment]
Yep, the terminology train left in March.

File: src/coins.cpp [hunk 1]
```diff
@@ -394,9 +394,9 @@ bool CCoinsViewCache::HavePourRequirements(const CTransaction& tx) const
 {
     boost::unordered_map<uint256, ZCIncrementalMerkleTree, CCoinsKeyHasher> intermediates;
 
-    BOOST_FOREACH(const CPourTx &pour, tx.vpour)
+    BOOST_FOREACH(const JSDescription &pour, tx.vjoinsplit)
```
Comment (daira) [inline src/coins.cpp]: Also change the local variable ``pour``.

File: src/coins.cpp [hunk 2]
```diff
@@ -394,9 +394,9 @@ bool CCoinsViewCache::HavePourRequirements(const CTransaction& tx) const
...
     {
-        BOOST_FOREACH(const uint256& serial, pour.serials)
+        BOOST_FOREACH(const uint256& serial, pour.nullifiers)
```
Comment (daira) [inline src/coins.cpp]: Also change ``serial``.

File: src/gtest/test_checktransaction.cpp [hunk 3]
```diff
@@ -15,9 +15,9 @@ TEST(checktransaction_tests, check_vpub_not_both_nonzero) {
         CValidationState state;
         state.SetPerformPourVerification(false); // don't verify the snark
```
Comment (daira) [inline src/gtest/test_checktransaction.cpp]: Change this method name.

File: src/gtest/test_checktransaction.cpp [hunk 4]
```diff
@@ -15,9 +15,9 @@ TEST(checktransaction_tests, check_vpub_not_both_nonzero) {
...
 
-        newTx.vpour.push_back(CPourTx());
+        newTx.vjoinsplit.push_back(JSDescription());
 
-        CPourTx *pourtx = &newTx.vpour[0];
+        JSDescription *pourtx = &newTx.vjoinsplit[0];
```
Comment (daira) [inline src/gtest/test_checktransaction.cpp]: Also change ``pourtx``. Etc. I know this is a WIP so I'll stop here.

File: src/main.cpp [hunk 5]
```diff
@@ -951,24 +951,24 @@ bool CheckTransaction(const CTransaction& tx, CValidationState &state)
...
     {
-        BOOST_FOREACH(const uint256& serial, pour.serials)
+        BOOST_FOREACH(const uint256& serial, pour.nullifiers)
         {
-            if (vPourSerials.count(serial))
-                return state.DoS(100, error("CheckTransaction(): duplicate serials"),
-                             REJECT_INVALID, "bad-pours-serials-duplicate");
+            if (vJoinSplitNullifiers.count(serial))
+                return state.DoS(100, error("CheckTransaction(): duplicate nullifiers"),
+                             REJECT_INVALID, "bad-pours-nullifiers-duplicate");
```
Comment (daira) [inline src/main.cpp]: "bad-joinsplit-nullifiers-duplicate"

**Comment by ebfull** [comment]
Sorry, this is a WIP because I have to do a lot more still.

**Comment by ebfull** [comment]
Rebased, working on the remaining bits..

**Comment by ebfull** [comment]
@zkbot try

**Comment by ebfull** [comment]
This is likely to fail the RPC tests because we haven't changed "encryptedBucket" yet.

**Comment by ebfull** [comment]
@zkbot try

**Comment by ebfull** [comment]
@zkbot r+

**Comment by ebfull** [comment]
@zkbot r-

**Comment by ebfull** [comment]
@zkbot try

**Comment by ebfull** [comment]
@zkbot try

**Comment by ebfull** [comment]
This is ready for review!

File: qa/rpc-tests/zcjoinsplitdoublespend.py [hunk 6]
```diff
@@ -56,57 +56,57 @@ def run_test(self):
...
 
-            receive_result = self.nodes[3].zcrawreceive(zcsecretkey, enc_bucket)
+            receive_result = self.nodes[3].zcrawreceive(zcsecretkey, enc_note)
             assert_equal(receive_result["exists"], True)
 
         blank_tx = self.nodes[0].createrawtransaction([], {})
         # Create pour {A, B}->{*}
-        pour_AB = self.nodes[0].zcrawpour(blank_tx,
+        pour_AB = self.nodes[0].zcrawjoinsplit(blank_tx,
                                           {pool[0] : zcsecretkey, pool[1] : zcsecretkey},
```
Comment (daira) [inline qa/rpc-tests/zcjoinsplitdoublespend.py]: Fix indentation.

File: src/test/coins_tests.cpp [hunk 7]
```diff
@@ -224,21 +224,21 @@ BOOST_AUTO_TEST_CASE(chained_pours)
 
     ZCIncrementalMerkleTree tree;
 
-    CPourTx ptx1;
+    JSDescription ptx1;
```
Comment (daira) [inline src/test/coins_tests.cpp]: I believe the "p" originally stood for "pour", so maybe change this to ``jstx1``? Low priority.

File: src/test/sighash_tests.cpp [hunk 8]
```diff
@@ -122,16 +122,16 @@ void static RandomTransaction(CMutableTransaction &tx, bool fSingle) {
     }
     if (tx.nVersion >= 2) {
         for (int pour = 0; pour < pours; pour++) {
-            CPourTx pourtx;
+            JSDescription pourtx;
```
Comment (daira) [inline src/test/sighash_tests.cpp]: ``s/pourtx/jstx/``

File: src/test/sighash_tests.cpp [hunk 9]
```diff
@@ -122,16 +122,16 @@ void static RandomTransaction(CMutableTransaction &tx, bool fSingle) {
     }
     if (tx.nVersion >= 2) {
         for (int pour = 0; pour < pours; pour++) {
```
Comment (daira) [inline src/test/sighash_tests.cpp]: ``s/pour/joinsplit/``

File: src/test/coins_tests.cpp (hunk 7 above)
Comment (daira) [inline src/test/coins_tests.cpp]: Also change the name of the test case (a few lines above this diff).

File: src/wallet/rpcwallet.cpp [hunk 10]
```diff
@@ -2644,17 +2644,17 @@ Value zc_raw_pour(const json_spirit::Array& params, bool fHelp)
     mtx.nVersion = 2;
     mtx.joinSplitPubKey = joinSplitPubKey;
 
-    CPourTx pourtx(*pzcashParams,
+    JSDescription jsdescription(*pzcashParams,
                    joinSplitPubKey,
```
Comment (daira) [inline src/wallet/rpcwallet.cpp]: Fix indentation. (And use ``jsdesc`` for consistency?)

File: src/zcbenchmarks.cpp [hunk 11]
```diff
@@ -76,7 +76,7 @@ double benchmark_create_joinsplit()
     uint256 anchor = ZCIncrementalMerkleTree().root();
 
     timer_start();
-    CPourTx pourtx(*pzcashParams,
+    JSDescription jsdescription(*pzcashParams,
                    pubKeyHash,
```
Comment (daira) [inline src/zcbenchmarks.cpp]: Fix indentation, use ``jsdesc``.

File: src/wallet/wallet.cpp [hunk 12]
```diff
@@ -1065,21 +1065,21 @@ void CWallet::WitnessBucketCommitment(std::vector<uint256> commitments,
 
         BOOST_FOREACH(const CTransaction& tx, block.vtx)
         {
-            BOOST_FOREACH(const CPourTx& pour, tx.vpour)
+            BOOST_FOREACH(const JSDescription& pour, tx.vjoinsplit)
```
Comment (daira) [inline src/wallet/wallet.cpp]: ``s/pour/jsdesc/``

File: src/test/coins_tests.cpp [hunk 13]
```diff
@@ -224,21 +224,21 @@ BOOST_AUTO_TEST_CASE(chained_pours)
...
     ptx1.anchor = tree.root();
     ptx1.commitments[0] = appendRandomCommitment(tree);
     ptx1.commitments[1] = appendRandomCommitment(tree);
 
     // Although it's not possible given our assumptions, if
     // two pours create the same treestate twice, we should
```
Comment (daira) [inline src/test/coins_tests.cpp]: s/pours/joinsplits/

**Comment by daira** [comment]
rpcclient.cpp has:
```
src/rpcclient.cpp:94:    { "zcrawpour", 1 },
src/rpcclient.cpp:95:    { "zcrawpour", 2 },
src/rpcclient.cpp:96:    { "zcrawpour", 3 },
src/rpcclient.cpp:97:    { "zcrawpour", 4 },
```

File: src/txdb.cpp [hunk 14]
```diff
@@ -81,7 +81,7 @@ bool CCoinsViewDB::GetAnchorAt(const uint256 &rt, ZCIncrementalMerkleTree &tree)
     return read;
 }
 
-bool CCoinsViewDB::GetSerial(const uint256 &serial) const {
+bool CCoinsViewDB::GetNullifier(const uint256 &serial) const {
```
Comment (daira) [inline src/txdb.cpp]: ``s/serial/nullifier`` (or ``nf``)

**Comment by daira** [comment]
In txdb.cpp, rename ``DB_SERIAL`` (and maybe change the character to ``'n'``), and ``BatchWriteSerial``.

File: src/txmempool.cpp [hunk 15]
```diff
@@ -99,9 +99,9 @@ bool CTxMemPool::addUnchecked(const uint256& hash, const CTxMemPoolEntry &entry,
     const CTransaction& tx = mapTx[hash].GetTx();
     for (unsigned int i = 0; i < tx.vin.size(); i++)
         mapNextTx[tx.vin[i].prevout] = CInPoint(&tx, i);
-    BOOST_FOREACH(const CPourTx &pour, tx.vpour) {
-        BOOST_FOREACH(const uint256 &serial, pour.serials) {
-            mapSerials[serial] = &tx;
+    BOOST_FOREACH(const JSDescription &joinsplit, tx.vjoinsplit) {
+        BOOST_FOREACH(const uint256 &serial, joinsplit.nullifiers) {
```
Comment (daira) [inline src/txmempool.cpp]: ``s/serial/nullifier/`` (or ``nf``)

File: src/txmempool.cpp [hunk 16]
```diff
@@ -148,9 +148,9 @@ void CTxMemPool::remove(const CTransaction &origTx, std::list<CTransaction>& rem
             }
             BOOST_FOREACH(const CTxIn& txin, tx.vin)
                 mapNextTx.erase(txin.prevout);
-            BOOST_FOREACH(const CPourTx& pour, tx.vpour) {
-                BOOST_FOREACH(const uint256& serial, pour.serials) {
-                    mapSerials.erase(serial);
+            BOOST_FOREACH(const JSDescription& joinsplit, tx.vjoinsplit) {
+                BOOST_FOREACH(const uint256& serial, joinsplit.nullifiers) {
```
Comment (daira) [inline src/txmempool.cpp]: ``s/serial/nullifier/`` (or ``nf``)

File: src/txmempool.cpp [hunk 17]
```diff
@@ -230,10 +230,10 @@ void CTxMemPool::removeConflicts(const CTransaction &tx, std::list<CTransaction>
         }
     }
 
-    BOOST_FOREACH(const CPourTx &pour, tx.vpour) {
-        BOOST_FOREACH(const uint256 &serial, pour.serials) {
-            std::map<uint256, const CTransaction*>::iterator it = mapSerials.find(serial);
-            if (it != mapSerials.end()) {
+    BOOST_FOREACH(const JSDescription &joinsplit, tx.vjoinsplit) {
+        BOOST_FOREACH(const uint256 &serial, joinsplit.nullifiers) {
```
Comment (daira) [inline src/txmempool.cpp]: ``s/serial/nullifier/`` (or ``nf``)

File: src/txmempool.cpp [hunk 18]
```diff
@@ -317,20 +317,20 @@ void CTxMemPool::check(const CCoinsViewCache *pcoins) const
 
         boost::unordered_map<uint256, ZCIncrementalMerkleTree, CCoinsKeyHasher> intermediates;
 
-        BOOST_FOREACH(const CPourTx &pour, tx.vpour) {
-            BOOST_FOREACH(const uint256 &serial, pour.serials) {
-                assert(!pcoins->GetSerial(serial));
+        BOOST_FOREACH(const JSDescription &joinsplit, tx.vjoinsplit) {
+            BOOST_FOREACH(const uint256 &serial, joinsplit.nullifiers) {
```
Comment (daira) [inline src/txmempool.cpp]: ``s/serial/nullifier/`` (or ``nf``)

File: src/txdb.h [hunk 19]
```diff
@@ -35,7 +35,7 @@ class CCoinsViewDB : public CCoinsView
     CCoinsViewDB(size_t nCacheSize, bool fMemory = false, bool fWipe = false);
 
     bool GetAnchorAt(const uint256 &rt, ZCIncrementalMerkleTree &tree) const;
-    bool GetSerial(const uint256 &serial) const;
+    bool GetNullifier(const uint256 &serial) const;
```
Comment (daira) [inline src/txdb.h]: ``s/serial/nullifier/``

File: src/rpcrawtransaction.cpp [hunk 20]
```diff
@@ -90,48 +90,48 @@ void TxToJSON(const CTransaction& tx, const uint256 hashBlock, Object& entry)
...
 
-        pour.push_back(Pair("anchor", pourtx.anchor.GetHex()));
+        joinsplit.push_back(Pair("anchor", jsdescription.anchor.GetHex()));
 
         {
-            Array serials;
-            BOOST_FOREACH(const uint256 serial, pourtx.serials) {
-                serials.push_back(serial.GetHex());
+            Array nullifiers;
+            BOOST_FOREACH(const uint256 serial, jsdescription.nullifiers) {
```
Comment (daira) [inline src/rpcrawtransaction.cpp]: ``s/serial/nullifier/`` (or ``nf``)

**Comment by daira** [comment]
ACK modulo remaining comments.

**Comment by daira** [comment]
And thanks for doing all this boring nitpicky renaming work!

**Comment by ebfull** [comment]
Ready to go, but I'd love if @defuse could take a look too. :)

File: qa/rpc-tests/zcjoinsplitdoublespend.py [hunk 21]
```diff
@@ -89,27 +89,27 @@ def run_test(self):
         blank_tx = self.nodes[0].createrawtransaction([], {})
         # Create pour {A, B}->{*}
         pour_AB = self.nodes[0].zcrawjoinsplit(blank_tx,
```
Comment (daira) [inline qa/rpc-tests/zcjoinsplitdoublespend.py]: s/pour/joinsplit/ here and in the comment.

(I think I missed all of the cases under qa/ in my previous review.)

**Comment by daira** [comment]
```
$ grep -Rn --exclude='*.o' --exclude='*.a' --exclude='*.Po' --exclude='*.ts' pour qa
qa/rpc-tests/zcjoinsplit.py:27:        pour_result = self.nodes[0].zcrawjoinsplit(protect_tx, {}, {zcaddress:39.9}, 39.9, 0)
qa/rpc-tests/zcjoinsplit.py:29:        receive_result = self.nodes[0].zcrawreceive(zcsecretkey, pour_result["encryptednote1"])
qa/rpc-tests/zcjoinsplit.py:32:        protect_tx = self.nodes[0].signrawtransaction(pour_result["rawtxn"])
qa/rpc-tests/zcjoinsplit.py:36:        receive_result = self.nodes[0].zcrawreceive(zcsecretkey, pour_result["encryptednote1"])
qa/rpc-tests/zcjoinsplit.py:39:        pour_tx = self.nodes[0].createrawtransaction([], {})
qa/rpc-tests/zcjoinsplit.py:40:        pour_result = self.nodes[0].zcrawjoinsplit(pour_tx, {receive_result["note"] : zcsecretkey}, {zcaddress: 39.8}, 0, 0.1)
qa/rpc-tests/zcjoinsplit.py:42:        self.nodes[0].sendrawtransaction(pour_result["rawtxn"])
qa/rpc-tests/zcjoinsplit.py:46:        receive_result = self.nodes[0].zcrawreceive(zcsecretkey, pour_result["encryptednote1"])
qa/rpc-tests/zcjoinsplitdoublespend.py:30:    def cannot_pour(self, node, txn):
qa/rpc-tests/zcjoinsplitdoublespend.py:40:    def expect_cannot_pour(self, node, txn):
qa/rpc-tests/zcjoinsplitdoublespend.py:41:        assert_equal(self.cannot_pour(node, txn), True)
qa/rpc-tests/zcjoinsplitdoublespend.py:90:        # Create pour {A, B}->{*}
qa/rpc-tests/zcjoinsplitdoublespend.py:91:        pour_AB = self.nodes[0].zcrawjoinsplit(blank_tx,
qa/rpc-tests/zcjoinsplitdoublespend.py:96:        # Create pour {B, C}->{*}
qa/rpc-tests/zcjoinsplitdoublespend.py:97:        pour_BC = self.nodes[0].zcrawjoinsplit(blank_tx,
qa/rpc-tests/zcjoinsplitdoublespend.py:102:        # Create pour {C, D}->{*}
qa/rpc-tests/zcjoinsplitdoublespend.py:103:        pour_CD = self.nodes[0].zcrawjoinsplit(blank_tx,
qa/rpc-tests/zcjoinsplitdoublespend.py:108:        # Create pour {A, D}->{*}
qa/rpc-tests/zcjoinsplitdoublespend.py:109:        pour_AD = self.nodes[0].zcrawjoinsplit(blank_tx,
qa/rpc-tests/zcjoinsplitdoublespend.py:114:        # (a)    Node 0 will spend pour AB, then attempt to
qa/rpc-tests/zcjoinsplitdoublespend.py:129:        AB_txid = self.nodes[0].sendrawtransaction(pour_AB["rawtxn"])
qa/rpc-tests/zcjoinsplitdoublespend.py:131:        self.expect_cannot_pour(self.nodes[0], pour_BC["rawtxn"])
qa/rpc-tests/zcjoinsplitdoublespend.py:142:        self.expect_cannot_pour(self.nodes[1], pour_BC["rawtxn"])
qa/rpc-tests/zcjoinsplitdoublespend.py:148:        self.expect_cannot_pour(self.nodes[0], pour_BC["rawtxn"])
qa/rpc-tests/zcjoinsplitdoublespend.py:149:        self.expect_cannot_pour(self.nodes[1], pour_BC["rawtxn"])
qa/rpc-tests/zcjoinsplitdoublespend.py:152:        self.nodes[2].sendrawtransaction(pour_BC["rawtxn"])
qa/rpc-tests/zcjoinsplitdoublespend.py:161:        self.expect_cannot_pour(self.nodes[0], pour_AB["rawtxn"])
qa/rpc-tests/zcjoinsplitdoublespend.py:162:        self.expect_cannot_pour(self.nodes[0], pour_CD["rawtxn"])
qa/rpc-tests/zcjoinsplitdoublespend.py:164:        self.expect_cannot_pour(self.nodes[1], pour_AB["rawtxn"])
qa/rpc-tests/zcjoinsplitdoublespend.py:165:        self.expect_cannot_pour(self.nodes[1], pour_CD["rawtxn"])
qa/rpc-tests/zcjoinsplitdoublespend.py:167:        self.expect_cannot_pour(self.nodes[2], pour_AB["rawtxn"])
qa/rpc-tests/zcjoinsplitdoublespend.py:168:        self.expect_cannot_pour(self.nodes[2], pour_CD["rawtxn"])
qa/rpc-tests/zcjoinsplitdoublespend.py:170:        self.expect_cannot_pour(self.nodes[3], pour_AB["rawtxn"])
qa/rpc-tests/zcjoinsplitdoublespend.py:171:        self.expect_cannot_pour(self.nodes[3], pour_CD["rawtxn"])
qa/rpc-tests/zcjoinsplitdoublespend.py:177:        self.nodes[0].sendrawtransaction(pour_AD["rawtxn"])
```

**Comment by ebfull** [comment]
Rebased.

**Comment by defuse** [comment]
ACK, changes look good and I couldn't find any unupdated terminology.

**Comment by ebfull** [comment]
@zkbot r+

**Comment by daira** [comment]
Post-hoc ACK for the changes after my last review.

**Pull request merged**